collector/infiniband_linux.go

```diff
@@ -0,0 +1,172 @@
+// Copyright 2017 The Prometheus Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// +build linux
+// +build !noinfiniband
+
+package collector
+
+import (
+	"errors"
+	"path/filepath"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/common/log"
+)
+
+const infinibandPath = "class/infiniband"
+
+var (
+	errInfinibandNoDevicesFound = errors.New("InfiniBand / No devices detected")
```

Comment:
One thing I forgot to mention last time (sorry!), error messages should have text that starts lowercase (so they can be chained if needed). You probably want some text like "no InfiniBand devices detected".  Same goes for the ports error.


An example of why it's typically done this way (with three different errors that each add context):

```
metrics collection failed: infiniband collector failed: no InfiniBand devices detected".
```